src/inmanta/deploy/scheduler.py

```diff
@@ -283,19 +297,31 @@ async def _new_version(
                 # => regularly pass control to the event loop to not block scheduler operation during update prep
                 await asyncio.sleep(0)
 
+            # A resource should not be present in more than one of these resource sets
+            assert len(new_desired_state | blocked_resources | unblocked_resources) == len(new_desired_state) + len(
+                blocked_resources
+            ) + len(unblocked_resources)
+
             # in the current implementation everything below the lock is synchronous, so it's not technically required. It is
             # however kept for two reasons:
             # 1. pass context once more to event loop before starting on the sync path
             #   (could be achieved with a simple sleep(0) if desired)
             # 2. clarity: it clearly signifies that this is the atomic and performance-sensitive part
             async with self._scheduler_lock:
                 self._state.version = version
+                for resource in blocked_resources:
+                    self._state.block_resource(resource, details, transient=False)
                 for resource in new_desired_state:
                     self._state.update_desired_state(resource, resources[resource])
                 for resource in added_requires.keys() | dropped_requires.keys():
                     self._state.update_requires(resource, requires[resource])
+                transitively_blocked_resources: set[ResourceIdStr] = self._state.block_provides(resources=blocked_resources)
+                for resource in unblocked_resources:
+                    self._state.unblock_resource(resource)
```

Comment:
What if we just marked it as transitively blocked? We'd unmark it here but that's not valid, is it?

Also potentially affects the `difference_update two lies down.